### PR TITLE
Vue 3 and Typescript support

### PR DESCRIPTION
**Pull request by shlroland**
I refactor the code to Vue3 and gen the d.ts to adapt typesscript

### CODE REVIEW

**Comment by gabriele-chiodi** [comment]
Hello, when will be released the version with support of Vue3 and Typescript?

File: package.json [hunk 1]
```diff
@@ -25,10 +26,17 @@
...
     "@rollup/plugin-commonjs": "^15.0.0",
-    "@rollup/plugin-node-resolve": "^9.0.0"
+    "@rollup/plugin-node-resolve": "^9.0.0",
+    "rollup": "^2.26.8",
+    "rollup-plugin-terser": "^7.0.2",
+    "rollup-plugin-typescript2": "^0.30.0",
+    "typescript": "^4.2.3",
+    "vue": "^3.0.9"
+  },
+  "peerDependencies": {
```
[the remark this answers is not in the document]
Comment (shlroland) [reply package.json:38]: I think it's better to create another branch for Vue3 other than the main branch.What do you think？

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    updateWatchVariables(true);
+  };
+  keycloak.onAuthRefreshError = function () {
+    updateWatchVariables(false);
+    typeof options.onAuthRefreshError === "function" &&
+      options.onAuthRefreshError(keycloak);
+  };
+  keycloak
+    .init(options.init)
+    .then((authenticated) => {
```
[the remark this answers is not in the document]
Comment (shlroland) [reply src/index.ts:112]: don't need Explicit. type inference is working.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        console.log(err);
+      });
+  },
+};
+
+function init(config: KConfig, watch: VueKeycloakInstance, options:VueKeycloakOptions) {
+  const ctor = sanitizeConfig(config);
+  const keycloak = Keycloak(ctor);
+
+  keycloak.onReady = function (authenticated) {
```
[the remark this answers is not in the document]
Comment (shlroland) [reply src/index.ts:80]: type inference is working.

File: src/types.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  KeycloakRoles,
+  KeycloakTokenParsed,
+} from "keycloak-js";
+
+export { KeycloakConfig, KeycloakError } from "keycloak-js";
+
+export type AnyFunction = (...args: any[]) => any;
+
+export interface KConfig extends KeycloakConfig {
+  authRealm?: string;
```
[the remark this answers is not in the document]
Comment (shlroland) [reply src/types.ts:19]: i think so

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+      refreshToken: null,
+      refreshTokenParsed: null,
+      timeSkew: null,
+      responseMode: null,
+      responseType: null,
+      hasRealmRole: null,
+      hasResourceRole: null,
+      keycloak: null,
+    });
+    Object.defineProperty(app.config.globalProperties, "$keycloak", {
```
[the remark this answers is not in the document]
Comment (shlroland) [reply src/index.ts]: i use  [vue-demi](https://github.com/vueuse/vue-demi) to replace vue

File: src/types.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (baltom) [reply src/types.ts:29]: I can see you have remove the `sanitizeConfig` stuff now. However the `authRealm`, `authUrl` & `authClientId` can also be removed As they will no longer get picked up by anything

File: package.json [hunk 6]
```diff
@@ -16,7 +17,8 @@
   },
   "homepage": "https://github.com/dsb-norge/vue-keycloak-js",
   "dependencies": {
-    "keycloak-js": "9.0.3"
+    "keycloak-js": "9.0.3",
+    "vue-demi": "latest"
```
Comment (baltom) [inline package.json:21]: Not quite sure that we need to introduce a new dependency for this support. What you had earlier worked fine in a Vue 2 project given the little check i mentioned in the other comment. While with the current way it crashes when attempting to reference "vue.reactive" when the package is built. 

Haven't tried for a Vue 3 project though, as i assume thats what you are testing it with. As mentioned in https://github.com/dsb-norge/vue-keycloak-js/issues/56 we have not yet started to use Vue 3.

**Pull request merged**